### PR TITLE
Fixing conversion from nested objects to url

### LINKED ISSUE #149: Enable to get commits of a branch
Currently, vsts-node-api can not get commits of a branch, while REST api has this feature https://www.visualstudio.com/en-us/docs/integrate/api/git/commits#in-the-history-of-a-branch.

### PR DESCRIPTION
**Pull request by damccorm**
Converting nested objects to url, while improved by #206, still wasn't quite right. Looks like this was also the underlying issue for at least issue #149, investigating to see if its caused any more known issues.

Fixes #149

### CODE REVIEW

File: api/VsoClient.ts [hunk 1]
```diff
@@ -183,41 +183,38 @@ export class VsoClient {
...
+    private queryParamsToString(queryParams: any): string {
+        let queryString: string = '?' + this.queryParamsToStringHelper(queryParams, '');
+
+        //Will always need to slice either a ? or & off of the end
+        return queryString.slice(0,-1);
     }
 
     protected getRequestUrl(routeTemplate: string, area: string, resource: string, routeValues: any, queryParams?: any): string {
-
+        console.log("PARAMS", queryParams);
```
Comment (stephenmichaelf) [inline api/VsoClient.ts]: Remove.

File: api/VsoClient.ts [hunk 2]
```diff
@@ -231,20 +228,10 @@ export class VsoClient {
...
-                    first = false;
-                } else {
-                    relativeUrl += "&" + valueString;
-                }
-            }
+        if (queryParams) {
+            relativeUrl += this.queryParamsToString(queryParams);
         }
-
+        console.log('Out', url.resolve(this.baseUrl, path.join(this.basePath, relativeUrl)));
```
Comment (stephenmichaelf) [inline api/VsoClient.ts]: Remove.

File: api/VsoClient.ts [hunk 3]
```diff
@@ -183,41 +183,38 @@ export class VsoClient {
...
-    private getValueString(queryValue, value) {
-        let valueString = null;
-        if (typeof(value) === 'object') {
-            valueString = this.getSerializedObject(queryValue, value);
-        } else {
-            valueString = queryValue + "=" + encodeURIComponent(value);
-        }
-        return valueString;
+    private queryParamsToString(queryParams: any): string {
+        let queryString: string = '?' + this.queryParamsToStringHelper(queryParams, '');
```
Comment (stephenmichaelf) [inline api/VsoClient.ts]: const

File: api/VsoClient.ts [hunk 4]
```diff
@@ -183,41 +183,38 @@ export class VsoClient {
...
+
+        //Will always need to slice either a ? or & off of the end
```
Comment (stephenmichaelf) [inline api/VsoClient.ts]: Nit: space between slashes and start of comment

File: api/VsoClient.ts [hunk 5]
```diff
@@ -183,41 +183,38 @@ export class VsoClient {
...
+        return queryString.slice(0,-1);
```
Comment (stephenmichaelf) [inline api/VsoClient.ts]: Why -1? How is this different than substring?

File: api/VsoClient.ts [hunk 6]
```diff
@@ -183,41 +183,38 @@ export class VsoClient {
...
+                    let newPrefix = prefix + encodeURIComponent(property.toString()) + '.';
+                    queryString += this.queryParamsToStringHelper(prop, newPrefix);
                 }
             }
         }
 
-        if (value == ""){
-            value += queryValue + "=" + object.toString();
+        if(queryString == '' && prefix.length > 0){
+            //Will always need to chop period off of end of prefix
```
Comment (stephenmichaelf) [inline api/VsoClient.ts]: Nit: space before slashes and comment.

File: api/VsoClient.ts [hunk 7]
```diff
@@ -183,41 +183,38 @@ export class VsoClient {
...
-            if (object.hasOwnProperty(property)) {
-                let prop = object[property];
-                let valueString = this.getValueString(property, prop);
-                if (first && prop !== undefined) {
-                    value += valueString;
-                    first = false;
-                } else if (prop !== undefined) {
-                    value += "&" + valueString;
+    private queryParamsToStringHelper(queryParams: any, prefix: string): string {
+        if (queryParams == undefined) {
```
Comment (stephenmichaelf) [inline api/VsoClient.ts]: ===

File: api/VsoClient.ts [hunk 8]
```diff
@@ -183,41 +183,38 @@ export class VsoClient {
...
+            return '';
+        }
+        let queryString: string = '';
+
+        if(typeof(queryParams) != 'string'){
```
Comment (stephenmichaelf) [inline api/VsoClient.ts]: !===

File: api/VsoClient.ts [hunk 9]
```diff
@@ -183,41 +183,38 @@ export class VsoClient {
...
+            for (let property in queryParams) {
+                if (queryParams.hasOwnProperty(property)){
+                    let prop = queryParams[property];
```
Comment (stephenmichaelf) [inline api/VsoClient.ts]: const

File: api/VsoClient.ts [hunk 10]
```diff
@@ -183,41 +183,38 @@ export class VsoClient {
...
+                    let newPrefix = prefix + encodeURIComponent(property.toString()) + '.';
```
Comment (stephenmichaelf) [inline api/VsoClient.ts]: const

File: api/VsoClient.ts [hunk 11]
```diff
@@ -183,41 +183,38 @@ export class VsoClient {
...
-        for (let property in object) {
-            if (object.hasOwnProperty(property)) {
-                let prop = object[property];
-                let valueString = this.getValueString(property, prop);
-                if (first && prop !== undefined) {
-                    value += valueString;
-                    first = false;
-                } else if (prop !== undefined) {
-                    value += "&" + valueString;
+    private queryParamsToStringHelper(queryParams: any, prefix: string): string {
```
Comment (stephenmichaelf) [inline api/VsoClient.ts]: Do you think we need to add more detailed tests for this functionality?

File: api/VsoClient.ts (hunk 5 above)
Comment (damccorm) [reply api/VsoClient.ts]: This is equivalent to substring(0, queryString.length - 1). Slice and substring are very similar, I tend to like slice in cases like this due to clarity but that's just a preference. https://stackoverflow.com/questions/2243824/what-is-the-difference-between-string-slice-and-string-substring

File: api/VsoClient.ts [hunk 12]
```diff
@@ -183,41 +183,38 @@ export class VsoClient {
...
+                    let prop = queryParams[property];
+                    let newPrefix = prefix + encodeURIComponent(property.toString()) + '.';
+                    queryString += this.queryParamsToStringHelper(prop, newPrefix);
                 }
             }
         }
 
-        if (value == ""){
-            value += queryValue + "=" + object.toString();
+        if(queryString == '' && prefix.length > 0){
```
Comment (stephenmichaelf) [inline api/VsoClient.ts]: ===

File: api/VsoClient.ts (hunk 11 above)
Comment (damccorm) [reply api/VsoClient.ts]: Good call, adding another test. This has caused issue so more coverage is definitely better.

**Comment by stephenmichaelf** [comment]
Bump patch version.

**Comment by stephenmichaelf** [comment]
Nit: For future branches it's best to name them users/damccorm/<short description> so we can keep track later if need be :)

File: api/VsoClient.ts [hunk 13]
```diff
@@ -183,41 +183,37 @@ export class VsoClient {
...
-                    first = false;
-                } else if (prop !== undefined) {
-                    value += "&" + valueString;
+    private queryParamsToStringHelper(queryParams: any, prefix: string): string {
+        if (queryParams === undefined) {
+            return '';
+        }
+        let queryString: string = '';
+
+        if(typeof(queryParams) !== 'string'){
```
Comment (stephenmichaelf) [inline api/VsoClient.ts]: Nit: space before brace.

File: api/VsoClient.ts [hunk 14]
```diff
@@ -183,41 +183,37 @@ export class VsoClient {
...
+            for (let property in queryParams) {
+                if (queryParams.hasOwnProperty(property)){
```
Comment (stephenmichaelf) [inline api/VsoClient.ts]: Nit: space before brace.

File: test/units/vsoClientTests.ts [hunk 15]
```diff
@@ -137,7 +137,28 @@ describe('VSOClient Units', function () {
...
+            .reply(200, {
+                value: [{id: 'testLocation', maxVersion: '1', releasedVersion: '1', routeTemplate: 'testTemplate', area: 'testArea5', resourceName: 'testName', resourceVersion: '1'}]
+        });
+
+        //Act
+        const queryParams = {status: {innerstatus: 2}, version: '1', nestedObject: {nestedField: 'value', innerNestedObject: {key: 'val2'}}};
+        const res: vsom.ClientVersioningData = await vsoClient.getVersioningData('1', 'testArea5', 'testLocation', {'testKey': 'testValue'}, queryParams);
+
+        //Assert
+        assert(res.apiVersion === '1');
```
Comment (stephenmichaelf) [inline test/units/vsoClientTests.ts]: Better to use assert.equal so we get better error messages when tests fail?

**Pull request merged**